File: frontend/services/expenses.ts
/**
 * Expenses API Service - Complete expense management operations
 */

import { apiGet, apiPost, apiPut, apiDelete } from "./api";

// ============================================
// Types
// ============================================

export type ExpenseStatus = "draft" | "submitted" | "pending" | "approved" | "rejected" | "returned" | "paid";
export type ExpenseType = "meal" | "transport" | "accommodation" | "supplies" | "communication" | "entertainment" | "travel" | "software" | "equipment" | "other";
export type PaymentMethod = "cash" | "credit_card" | "debit_card" | "bank_transfer" | "company_card" | "petty_cash" | "other";

export interface ExpenseItem {
    id: string;
    expense_id: string;
    date: string;
    expense_type: ExpenseType;
    category_id?: string;
    amount: number;
    currency: string;
    currency_rate: number;
    description?: string;
    vendor?: string;
    attachment_url?: string;
    receipt_path?: string;
    ocr_data?: Record<string, unknown>;
    created_at: string;
    updated_at: string;
}

export interface ExpenseApproval {
    id: string;
    expense_id: string;
    approver_id: string;
    approver_name?: string;
    level: number;
    status: "pending" | "approved" | "rejected";
    decision_at?: string;
    comments?: string;
    created_at: string;
}

export interface ExpenseAuditLog {
    id: string;
    expense_id: string;
    user_id: string;
    user_name?: string;
    action: string;
    old_values?: Record<string, unknown>;
    new_values?: Record<string, unknown>;
    created_at: string;
}

export interface Expense {
    id: string;
    user_id: string;
    user_name?: string;
    title: string;
    description?: string;
    project_id?: string;
    project_name?: string;
    cost_center_id?: string;
    cost_center_name?: string;
    total_amount: number;
    currency: string;
    vendor?: string;
    payment_method: PaymentMethod;
    status: ExpenseStatus;
    rejection_reason?: string;
    return_reason?: string;
    current_approval_level: number;
    required_approval_levels: number;
    created_at: string;
    updated_at: string;
    submitted_at?: string;
    approved_at?: string;
    rejected_at?: string;
    paid_at?: string;
    items?: ExpenseItem[];
    approvals?: ExpenseApproval[];
}

export interface ExpenseCreate {
    title: string;
    description?: string;
    project_id?: string;
    cost_center_id?: string;
    currency?: string;
    vendor?: string;
    payment_method?: PaymentMethod;
    items?: ExpenseItemCreate[];
}

export interface ExpenseUpdate {
    title?: string;
    description?: string;
    project_id?: string;
    cost_center_id?: string;
    currency?: string;
    vendor?: string;
    payment_method?: PaymentMethod;
}

export interface ExpenseItemCreate {
    date: string;
    expense_type?: ExpenseType;
    category_id?: string;
    amount: number;
    currency?: string;
    description?: string;
    vendor?: string;
}

// Dashboard types
export interface ExpenseDashboardStats {
    total_expenses: number;
    pending_count: number;
    approved_this_month: number;
    pending_approval_amount: number;
    my_expenses_count: number;
    my_pending_count: number;
}

export interface MonthlyTrend {
    month: string;
    year: number;
    amount: number;
    count: number;
}

export interface CategoryBreakdown {
    category: string;
    amount: number;
    count: number;
    percentage: number;
}

export interface DepartmentBreakdown {
    department: string;
    amount: number;
    count: number;
}

export interface ProjectBreakdown {
    project: string;
    project_id: string;
    amount: number;
    count: number;
}

export interface BudgetComparison {
    cost_center: string;
    budget: number;
    actual: number;
    variance: number;
    variance_percentage: number;
}

export interface ExpenseAnalytics {
    stats: {
        total_amount: number;
        total_count: number;
        pending_count: number;
        approved_count: number;
        rejected_count: number;
        average_amount: number;
    };
    monthly_trends: MonthlyTrend[];
    by_category: CategoryBreakdown[];
    by_department: DepartmentBreakdown[];
    by_project: ProjectBreakdown[];
    budget_comparison: BudgetComparison[];
}

export interface ExpenseCategory {
    id: string;
    name: string;
    description?: string;
    is_active: boolean;
}

export interface CostCenter {
    id: string;
    name: string;
    code: string;
    budget?: number;
}

// ============================================
// API Params
// ============================================

export interface ExpensesParams {
    skip?: number;
    limit?: number;
    status?: ExpenseStatus;
    project_id?: string;
    cost_center_id?: string;
    start_date?: string;
    end_date?: string;
    [key: string]: string | number | boolean | undefined;
}

// ============================================
// Expense CRUD API
// ============================================

const BASE_URL = "/expenses";

export async function getExpenses(params?: ExpensesParams): Promise<Expense[]> {
    return apiGet<Expense[]>(BASE_URL, params);
}

export async function getMyExpenses(params?: ExpensesParams): Promise<Expense[]> {
    return apiGet<Expense[]>(`${BASE_URL}/me`, params);
}

export async function getPendingExpenses(): Promise<Expense[]> {
    return apiGet<Expense[]>(`${BASE_URL}/pending`);
}

export async function getExpense(id: string): Promise<Expense> {
    return apiGet<Expense>(`${BASE_URL}/${id}`);
}

export async function createExpense(data: ExpenseCreate): Promise<Expense> {
    return apiPost<Expense>(BASE_URL, data);
}

export async function updateExpense(id: string, data: ExpenseUpdate): Promise<Expense> {
    return apiPut<Expense>(`${BASE_URL}/${id}`, data);
}

export async function deleteExpense(id: string): Promise<void> {
    return apiDelete(`${BASE_URL}/${id}`);
}

// ============================================
// Expense Workflow API
// ============================================

export async function submitExpense(id: string): Promise<Expense> {
    return apiPost<Expense>(`${BASE_URL}/${id}/submit`, {});
}

export async function approveExpense(id: string, comments?: string): Promise<Expense> {
    return apiPost<Expense>(`${BASE_URL}/${id}/approve`, { comments });
}

export async function rejectExpense(id: string, reason: string, comments?: string): Promise<Expense> {
    return apiPost<Expense>(`${BASE_URL}/${id}/reject`, { reason, comments });
}

export async function returnExpense(id: string, reason: string): Promise<Expense> {
    return apiPost<Expense>(`${BASE_URL}/${id}/return`, { reason });
}

export async function markExpensePaid(id: string): Promise<Expense> {
    return apiPost<Expense>(`${BASE_URL}/${id}/pay`, {});
}

// ============================================
// Expense Items API
// ============================================

export async function addExpenseItem(expenseId: string, item: ExpenseItemCreate): Promise<ExpenseItem> {
    return apiPost<ExpenseItem>(`${BASE_URL}/${expenseId}/items`, item);
}

export async function updateExpenseItem(expenseId: string, itemId: string, item: Partial<ExpenseItemCreate>): Promise<ExpenseItem> {
    return apiPut<ExpenseItem>(`${BASE_URL}/${expenseId}/items/${itemId}`, item);
}

export async function deleteExpenseItem(expenseId: string, itemId: string): Promise<void> {
    return apiDelete(`${BASE_URL}/${expenseId}/items/${itemId}`);
}

// ============================================
// Receipt Upload API
// ============================================

export async function uploadReceipt(expenseId: string, file: File, itemId?: string): Promise<{ url: string; ocr_data?: Record<string, unknown> }> {
    const formData = new FormData();
    formData.append("file", file);
    if (itemId) {
        formData.append("item_id", itemId);
    }

    const response = await fetch(`/api${BASE_URL}/${expenseId}/receipt`, {
        method: "POST",
        body: formData,
        credentials: "include",
    });

    if (!response.ok) {
        throw new Error("Failed to upload receipt");
    }

    return response.json();
}

// ============================================
// Approval & Audit API
// ============================================

export async function getExpenseApprovals(expenseId: string): Promise<ExpenseApproval[]> {
    return apiGet<ExpenseApproval[]>(`${BASE_URL}/${expenseId}/approvals`);
}

export async function getExpenseAuditLog(expenseId: string): Promise<ExpenseAuditLog[]> {
    return apiGet<ExpenseAuditLog[]>(`${BASE_URL}/${expenseId}/audit-log`);
}

// ============================================
// Dashboard & Analytics API
// ============================================

const DASHBOARD_URL = "/expense-dashboard";

export async function getExpenseDashboardStats(): Promise<ExpenseDashboardStats> {
    return apiGet<ExpenseDashboardStats>(`${DASHBOARD_URL}/stats`);
}

export async function getExpenseAnalytics(params?: {
    start_date?: string;
    end_date?: string;
    year?: number;
}): Promise<ExpenseAnalytics> {
    return apiGet<ExpenseAnalytics>(`${DASHBOARD_URL}/analytics`, params);
}

export async function getMonthlyTrends(year?: number): Promise<{ trends: MonthlyTrend[] }> {
    return apiGet<{ trends: MonthlyTrend[] }>(`${DASHBOARD_URL}/monthly-trends`, year ? { year } : undefined);
}

export async function getCategoryBreakdown(params?: {
    start_date?: string;
    end_date?: string;
}): Promise<{ categories: CategoryBreakdown[] }> {
    return apiGet<{ categories: CategoryBreakdown[] }>(`${DASHBOARD_URL}/by-category`, params);
}

export async function getDepartmentBreakdown(params?: {
    start_date?: string;
    end_date?: string;
}): Promise<{ departments: DepartmentBreakdown[] }> {
    return apiGet<{ departments: DepartmentBreakdown[] }>(`${DASHBOARD_URL}/by-department`, params);
}

export async function getProjectBreakdown(params?: {
    start_date?: string;
    end_date?: string;
}): Promise<{ projects: ProjectBreakdown[] }> {
    return apiGet<{ projects: ProjectBreakdown[] }>(`${DASHBOARD_URL}/by-project`, params);
}

export async function getBudgetComparison(period: "monthly" | "quarterly" | "yearly" = "monthly"): Promise<{ comparisons: BudgetComparison[] }> {
    return apiGet<{ comparisons: BudgetComparison[] }>(`${DASHBOARD_URL}/budget-comparison`, { period });
}

// ============================================
// Reports API
// ============================================

const REPORTS_URL = "/expense-reports";

export async function exportExpenseReport(params: {
    format: "pdf" | "excel";
    start_date?: string;
    end_date?: string;
    user_id?: string;
    department_id?: string;
    status?: ExpenseStatus;
}): Promise<Blob> {
    const queryParams = new URLSearchParams();
    Object.entries(params).forEach(([key, value]) => {
        if (value) queryParams.append(key, value);
    });

    const response = await fetch(`/api${REPORTS_URL}/export?${queryParams}`, {
        method: "GET",
        credentials: "include",
    });

    if (!response.ok) {
        throw new Error("Failed to export report");
    }

    return response.blob();
}

export async function getTaxReport(params: {
    year: number;
    quarter?: number;
}): Promise<Blob> {
    const queryParams = new URLSearchParams();
    queryParams.append("year", params.year.toString());
    if (params.quarter) queryParams.append("quarter", params.quarter.toString());

    const response = await fetch(`/api${REPORTS_URL}/tax?${queryParams}`, {
        method: "GET",
        credentials: "include",
    });

    if (!response.ok) {
        throw new Error("Failed to generate tax report");
    }

    return response.blob();
}

// ============================================
// Categories & Cost Centers API
// ============================================

export async function getExpenseCategories(): Promise<ExpenseCategory[]> {
    return apiGet<ExpenseCategory[]>("/expense-categories");
}

export async function getCostCenters(): Promise<CostCenter[]> {
    return apiGet<CostCenter[]>("/cost-centers");
}

// ============================================
// Helper Functions
// ============================================

export function getStatusColor(status: ExpenseStatus): string {
    const colors: Record<ExpenseStatus, string> = {
        draft: "bg-gray-500/20 text-gray-400",
        submitted: "bg-blue-500/20 text-blue-400",
        pending: "bg-yellow-500/20 text-yellow-400",
        approved: "bg-green-500/20 text-green-400",
        rejected: "bg-red-500/20 text-red-400",
        returned: "bg-orange-500/20 text-orange-400",
        paid: "bg-emerald-500/20 text-emerald-400",
    };
    return colors[status] || "bg-gray-500/20 text-gray-400";
}

export function getStatusLabel(status: ExpenseStatus): string {
    const labels: Record<ExpenseStatus, string> = {
        draft: "Draft",
        submitted: "Submitted",
        pending: "Pending Approval",
        approved: "Approved",
        rejected: "Rejected",
        returned: "Returned",
        paid: "Paid",
    };
    return labels[status] || status;
}

export function formatCurrency(amount: number, currency: string = "EGP"): string {
    return new Intl.NumberFormat("en-US", {
        style: "currency",
        currency: currency,
        minimumFractionDigits: 2,
    }).format(amount);
}

export function getExpenseTypeLabel(type: ExpenseType): string {
    const labels: Record<ExpenseType, string> = {
        meal: "Meal",
        transport: "Transport",
        accommodation: "Accommodation",
        supplies: "Supplies",
        communication: "Communication",
        entertainment: "Entertainment",
        travel: "Travel",
        software: "Software",
        equipment: "Equipment",
        other: "Other",
    };
    return labels[type] || type;
}

export function getPaymentMethodLabel(method: PaymentMethod): string {
    const labels: Record<PaymentMethod, string> = {
        cash: "Cash",
        credit_card: "Credit Card",
        debit_card: "Debit Card",
        bank_transfer: "Bank Transfer",
        company_card: "Company Card",
        petty_cash: "Petty Cash",
        other: "Other",
    };
    return labels[method] || method;
}

File: backend/app/routers/expenses.py
"""Expenses Router - CRUD and workflow operations for expenses."""
from typing import List, Optional
from datetime import datetime
from decimal import Decimal
from fastapi import APIRouter, Depends, HTTPException, status, UploadFile, File, Form
from sqlalchemy.orm import Session
from app.database import get_db
from app.models import (
    Expense, ExpenseItem, ExpenseApproval, ExpenseAuditLog, 
    User, ExpenseStatus, ApprovalStatus
)
from app.schemas import (
    ExpenseCreate, ExpenseUpdate, ExpenseResponse,
    ExpenseRejectAction, ExpenseReturnAction, ExpenseApprovalResponse,
    ExpenseAuditLogResponse
)
from app.utils import get_current_active_user
from app.services.file_upload import save_receipt
from app.services.notification_service import NotificationService

router = APIRouter()


def create_audit_log(
    db: Session,
    expense_id: str,
    user_id: str,
    action: str,
    old_values: dict = None,
    new_values: dict = None
):
    """Create an audit log entry."""
    log = ExpenseAuditLog(
        expense_id=expense_id,
        user_id=user_id,
        action=action,
        old_values=old_values,
        new_values=new_values
    )
    db.add(log)


@router.get("/", response_model=List[ExpenseResponse])
def get_all_expenses(
    skip: int = 0,
    limit: int = 100,
    user_id: Optional[str] = None,
    status_filter: Optional[str] = None,
    project_id: Optional[str] = None,
    cost_center_id: Optional[str] = None,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_active_user)
):
    """Get all expenses with optional filters."""
    query = db.query(Expense)
    
    # Non-admins see their own or pending for approval
    if current_user.role not in ["admin", "manager"]:
        query = query.filter(Expense.user_id == current_user.id)
    elif user_id:
        query = query.filter(Expense.user_id == user_id)
    
    if status_filter:
        query = query.filter(Expense.status == status_filter)
    if project_id:
        query = query.filter(Expense.project_id == project_id)
    if cost_center_id:
        query = query.filter(Expense.cost_center_id == cost_center_id)
    
    expenses = query.order_by(Expense.created_at.desc()).offset(skip).limit(limit).all()
    
    # Populate user info for each expense
    for expense in expenses:
        expense.user = db.query(User).filter(User.id == expense.user_id).first()
    
    return expenses


@router.get("/my", response_model=List[ExpenseResponse])
def get_my_expenses(
    status_filter: Optional[str] = None,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_active_user)
):
    """Get current user's expenses."""
    query = db.query(Expense).filter(Expense.user_id == current_user.id)
    
    if status_filter:
        query = query.filter(Expense.status == status_filter)
    
    return query.order_by(Expense.created_at.desc()).all()


@router.get("/pending", response_model=List[ExpenseResponse])
def get_pending_expenses(
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_active_user)
):
    """Get pending expenses for approval (managers only)."""
    if current_user.role not in ["admin", "manager"]:
        raise HTTPException(status_code=403, detail="Manager access required")
    
    expenses = db.query(Expense).filter(
        Expense.status.in_([ExpenseStatus.PENDING.value, ExpenseStatus.SUBMITTED.value])
    ).order_by(Expense.created_at.desc()).all()
    
    # Populate user info
    for expense in expenses:
        expense.user = db.query(User).filter(User.id == expense.user_id).first()
    
    return expenses


@router.post("/", response_model=ExpenseResponse, status_code=status.HTTP_201_CREATED)
def create_expense(
    expense_data: ExpenseCreate,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_active_user)
):
    """Create a new expense."""
    db_expense = Expense(
        user_id=current_user.id,
        title=expense_data.title,
        description=expense_data.description,
        project_id=expense_data.project_id,
        cost_center_id=expense_data.cost_center_id,
        currency=expense_data.currency,
        vendor=expense_data.vendor,
        payment_method=expense_data.payment_method
    )
    
    db.add(db_expense)
    db.flush()
    
    # Add items if provided
    total_amount = Decimal("0.00")
    if expense_data.items:
        for item_data in expense_data.items:
            item = ExpenseItem(
                expense_id=db_expense.id,
                **item_data.model_dump()
            )
            db.add(item)
            # Calculate with currency rate
            total_amount += Decimal(str(item_data.amount)) * Decimal(str(item_data.currency_rate))
    
    db_expense.total_amount = total_amount
    
    # Create audit log
    create_audit_log(
        db, db_expense.id, current_user.id, "created",
        new_values={"title": expense_data.title, "total_amount": float(total_amount)}
    )
    
    db.commit()
    db.refresh(db_expense)
    return db_expense


@router.get("/{expense_id}", response_model=ExpenseResponse)
def get_expense(
    expense_id: str,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_active_user)
):
    """Get a specific expense by ID."""
    expense = db.query(Expense).filter(Expense.id == expense_id).first()
    if not expense:
        raise HTTPException(status_code=404, detail="Expense not found")
    
    # Check access
    if current_user.role not in ["admin", "manager"] and expense.user_id != current_user.id:
        raise HTTPException(status_code=403, detail="Not authorized")
    
    expense.user = db.query(User).filter(User.id == expense.user_id).first()
    return expense


@router.put("/{expense_id}", response_model=ExpenseResponse)
def update_expense(
    expense_id: str,
    expense_data: ExpenseUpdate,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_active_user)
):
    """Update an expense."""
    expense = db.query(Expense).filter(Expense.id == expense_id).first()
    if not expense:
        raise HTTPException(status_code=404, detail="Expense not found")
    
    # Check authorization
    if expense.user_id != current_user.id and current_user.role not in ["admin", "manager"]:
        raise HTTPException(status_code=403, detail="Not authorized")
    
    # Store old values for audit
    old_values = {"status": expense.status, "total_amount": float(expense.total_amount or 0)}
    
    # Update fields
    update_data = expense_data.model_dump(exclude_unset=True, exclude={"items", "status"})
    for key, value in update_data.items():
        setattr(expense, key, value)
    
    if expense_data.status is not None:
        if expense_data.status == "approved":
            expense.approved_at = datetime.utcnow()
        elif expense_data.status == "pending":
            expense.submitted_at = datetime.utcnow()
        elif expense_data.status == "rejected":
            expense.rejected_at = datetime.utcnow()
        elif expense_data.status == "paid":
            expense.paid_at = datetime.utcnow()
        expense.status = expense_data.status
    
    # Update items if provided
    if expense_data.items is not None:
        db.query(ExpenseItem).filter(ExpenseItem.expense_id == expense_id).delete()
        
        total_amount = Decimal("0.00")
        for item_data in expense_data.items:
            item = ExpenseItem(
                expense_id=expense.id,
                **item_data.model_dump()
            )
            db.add(item)
            total_amount += Decimal(str(item_data.amount)) * Decimal(str(item_data.currency_rate))
        
        expense.total_amount = total_amount
    
    # Create audit log
    create_audit_log(
        db, expense.id, current_user.id, "updated",
        old_values=old_values,
        new_values={"status": expense.status, "total_amount": float(expense.total_amount or 0)}
    )
    
    db.commit()
    db.refresh(expense)
    return expense


@router.post("/{expense_id}/submit", response_model=ExpenseResponse)
def submit_expense(
    expense_id: str,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_active_user)
):
    """Submit an expense for approval."""
    expense = db.query(Expense).filter(Expense.id == expense_id).first()
    if not expense:
        raise HTTPException(status_code=404, detail="Expense not found")
    
    if expense.user_id != current_user.id:
        raise HTTPException(status_code=403, detail="Not authorized")
    
    if expense.status not in [ExpenseStatus.DRAFT.value, ExpenseStatus.RETURNED.value]:
        raise HTTPException(status_code=400, detail="Can only submit draft or returned expenses")
    
    old_status = expense.status
    expense.status = ExpenseStatus.SUBMITTED.value
    expense.submitted_at = datetime.utcnow()
    expense.current_approval_level = 1
    
    # Create audit log
    create_audit_log(
        db, expense.id, current_user.id, "submitted",
        old_values={"status": old_status},
        new_values={"status": expense.status}
    )
    
    db.commit()
    db.refresh(expense)
    return expense


@router.post("/{expense_id}/approve", response_model=ExpenseResponse)
def approve_expense(
    expense_id: str,
    comments: Optional[str] = None,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_active_user)
):
    """Approve an expense (managers only)."""
    if current_user.role not in ["admin", "manager"]:
        raise HTTPException(status_code=403, detail="Manager access required")
    
    expense = db.query(Expense).filter(Expense.id == expense_id).first()
    if not expense:
        raise HTTPException(status_code=404, detail="Expense not found")
    
    if expense.status not in [ExpenseStatus.SUBMITTED.value, ExpenseStatus.PENDING.value]:
        raise HTTPException(status_code=400, detail="Expense is not pending approval")
    
    # Create approval record
    approval = ExpenseApproval(
        expense_id=expense.id,
        approver_id=current_user.id,
        level=expense.current_approval_level,
        status=ApprovalStatus.APPROVED.value,
        decision_at=datetime.utcnow(),
        comments=comments
    )
    db.add(approval)
    
    # Check if more approvals needed
    if expense.current_approval_level >= expense.required_approval_levels:
        expense.status = ExpenseStatus.APPROVED.value
        expense.approved_at = datetime.utcnow()
    else:
        expense.current_approval_level += 1
        expense.status = ExpenseStatus.PENDING.value
    
    # Create audit log
    create_audit_log(
        db, expense.id, current_user.id, "approved",
        new_values={"status": expense.status, "level": expense.current_approval_level}
    )
    
    db.commit()
    db.refresh(expense)
    return expense


@router.post("/{expense_id}/reject", response_model=ExpenseResponse)
def reject_expense(
    expense_id: str,
    rejection: ExpenseRejectAction,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_active_user)
):
    """Reject an expense with reason (managers only)."""
    if current_user.role not in ["admin", "manager"]:
        raise HTTPException(status_code=403, detail="Manager access required")
    
    expense = db.query(Expense).filter(Expense.id == expense_id).first()
    if not expense:
        raise HTTPException(status_code=404, detail="Expense not found")
    
    old_status = expense.status
    expense.status = ExpenseStatus.REJECTED.value
    expense.rejection_reason = rejection.reason
    expense.rejected_at = datetime.utcnow()
    
    # Create approval record
    approval = ExpenseApproval(
        expense_id=expense.id,
        approver_id=current_user.id,
        level=expense.current_approval_level,
        status=ApprovalStatus.REJECTED.value,
        decision_at=datetime.utcnow(),
        comments=rejection.comments or rejection.reason
    )
    db.add(approval)
    
    reject_audit = create_audit_log(
        db=db,
        expense_id=expense.id,
        user_id=current_user.id,
        action="rejected",
        new_values={"status": ExpenseStatus.REJECTED.value, "rejection_reason": rejection.reason}
    )
    db.add(reject_audit)
    
    db.commit()
    db.refresh(expense)
    
    # Notify the user
    NotificationService.notify_expense_status(
        db=db,
        user_id=expense.user_id,
        expense_title=expense.description or "Expense",
        status="rejected",
        expense_id=expense.id
    )
    
    return expense


@router.post("/{expense_id}/return", response_model=ExpenseResponse)
def return_expense(
    expense_id: str,
    return_data: ExpenseReturnAction,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_active_user)
):
    """Return an expense for revision (managers only)."""
    if current_user.role not in ["admin", "manager"]:
        raise HTTPException(status_code=403, detail="Manager access required")
    
    expense = db.query(Expense).filter(Expense.id == expense_id).first()
    if not expense:
        raise HTTPException(status_code=404, detail="Expense not found")
    
    old_status = expense.status
    expense.status = ExpenseStatus.RETURNED.value
    expense.return_reason = return_data.reason
    
    # Create audit log
    create_audit_log(
        db, expense.id, current_user.id, "returned",
        old_values={"status": old_status},
        new_values={"status": expense.status, "reason": return_data.reason}
    )
    
    db.commit()
    db.refresh(expense)
    return expense


@router.post("/{expense_id}/mark-paid", response_model=ExpenseResponse)
def mark_expense_paid(
    expense_id: str,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_active_user)
):
    """Mark an approved expense as paid (managers only)."""
    if current_user.role not in ["admin", "manager"]:
        raise HTTPException(status_code=403, detail="Manager access required")
    
    expense = db.query(Expense).filter(Expense.id == expense_id).first()
    if not expense:
        raise HTTPException(status_code=404, detail="Expense not found")
    
    if expense.status != ExpenseStatus.APPROVED.value:
        raise HTTPException(status_code=400, detail="Only approved expenses can be marked as paid")
    
    old_status = expense.status
    expense.status = ExpenseStatus.PAID.value
    expense.paid_at = datetime.utcnow()
    
    # Create audit log
    create_audit_log(
        db, expense.id, current_user.id, "paid",
        old_values={"status": old_status},
        new_values={"status": expense.status}
    )
    
    db.commit()
    db.refresh(expense)
    return expense


@router.post("/{expense_id}/upload-receipt")
async def upload_receipt(
    expense_id: str,
    file: UploadFile = File(...),
    item_id: Optional[str] = Form(None),
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_active_user)
):
    """Upload a receipt for an expense item."""
    expense = db.query(Expense).filter(Expense.id == expense_id).first()
    if not expense:
        raise HTTPException(status_code=404, detail="Expense not found")
    
    if expense.user_id != current_user.id and current_user.role not in ["admin", "manager"]:
        raise HTTPException(status_code=403, detail="Not authorized")
    
    # Save the file
    file_path = await save_receipt(file, expense_id, item_id)
    
    # Update expense item if specified
    if item_id:
        item = db.query(ExpenseItem).filter(
            ExpenseItem.id == item_id,
            ExpenseItem.expense_id == expense_id
        ).first()
        if item:
            item.receipt_path = file_path
            db.commit()
    
    return {"message": "Receipt uploaded successfully", "path": file_path}


@router.get("/{expense_id}/approvals", response_model=List[ExpenseApprovalResponse])
def get_expense_approvals(
    expense_id: str,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_active_user)
):
    """Get approval history for an expense."""
    expense = db.query(Expense).filter(Expense.id == expense_id).first()
    if not expense:
        raise HTTPException(status_code=404, detail="Expense not found")
    
    if current_user.role not in ["admin", "manager"] and expense.user_id != current_user.id:
        raise HTTPException(status_code=403, detail="Not authorized")
    
    approvals = db.query(ExpenseApproval).filter(
        ExpenseApproval.expense_id == expense_id
    ).order_by(ExpenseApproval.level).all()
    
    result = []
    for approval in approvals:
        approver = db.query(User).filter(User.id == approval.approver_id).first()
        result.append(ExpenseApprovalResponse(
            id=approval.id,
            expense_id=approval.expense_id,
            approver_id=approval.approver_id,
            approver_name=approver.full_name if approver else None,
            level=approval.level,
            status=approval.status,
            decision_at=approval.decision_at,
            comments=approval.comments,
            created_at=approval.created_at
        ))
    
    return result


@router.get("/{expense_id}/audit-log", response_model=List[ExpenseAuditLogResponse])
def get_expense_audit_log(
    expense_id: str,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_active_user)
):
    """Get audit log for an expense."""
    expense = db.query(Expense).filter(Expense.id == expense_id).first()
    if not expense:
        raise HTTPException(status_code=404, detail="Expense not found")
    
    if current_user.role not in ["admin", "manager"] and expense.user_id != current_user.id:
        raise HTTPException(status_code=403, detail="Not authorized")
    
    logs = db.query(ExpenseAuditLog).filter(
        ExpenseAuditLog.expense_id == expense_id
    ).order_by(ExpenseAuditLog.created_at.desc()).all()
    
    result = []
    for log in logs:
        user = db.query(User).filter(User.id == log.user_id).first()
        result.append(ExpenseAuditLogResponse(
            id=log.id,
            expense_id=log.expense_id,
            user_id=log.user_id,
            user_name=user.full_name if user else None,
            action=log.action,
            old_values=log.old_values,
            new_values=log.new_values,
            created_at=log.created_at
        ))
    
    return result


@router.delete("/{expense_id}", status_code=status.HTTP_204_NO_CONTENT)
def delete_expense(
    expense_id: str,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_active_user)
):
    """Delete an expense (draft only)."""
    expense = db.query(Expense).filter(Expense.id == expense_id).first()
    if not expense:
        raise HTTPException(status_code=404, detail="Expense not found")
    
    if expense.user_id != current_user.id and current_user.role != "admin":
        raise HTTPException(status_code=403, detail="Not authorized")
    
    if expense.status != ExpenseStatus.DRAFT.value:
        raise HTTPException(status_code=400, detail="Can only delete draft expenses")
    
    db.delete(expense)
    db.commit()
    return None
